### DISCUSSION TITLE
INSFV model - periodic boundary conditions

### DISCUSSION THREAD

**Discussion by thomas-petersen-1** [idaholab/moose#25946] [Q&A General]
Does MOOSE currently support periodic boundary conditions in its finite volume implementation of the Navier-Stokes equations?

**Comment by GiudGiud** [comment]
Hello

For velocity?
I ll check tomorrow what we need for that. We have not looked into it so it s unlikely we support it

Guillaume

**Reply by thomas-petersen-1** [reply]
Yes, for the velocity components. Thank you for verifying, Guillaume.

**Reply by GiudGiud** [reply]
Periodic BCs seem to be silently ignored. 

If I set a flat inlet velocity it gets obeyed first.
If I set an inlet and outlet pressure, then the periodic solution in velocity appears with or without PeriodicBCs (could you use that?)
If I set an outlet pressure and a momemtum source and PeriodicBCs I dont get convergence which makes me think they are not applied

**Reply by GiudGiud** [reply]
@roystgnr do we do periodic BCs on constant monomials?

**Reply by grmnptr** [reply]
I dont think we have an FVPeriodicBC. FV has its own system for BCs.

**Reply by grmnptr** [reply]
If I remember correctly in the assembly loops we query FVFluxBCs and the FVDirichletBCs are applied in the different residual terms which need face values (laplacians, gradients etc).

**Reply by GiudGiud** [reply]
If we did it on the MOOSE side it would be a dirichlet BC on one side that goes to retrieve the value on the other side. 

Periodic BCs are not done in MOOSE, they rely on some libmesh mechanics I havent dwelled in

**Reply by grmnptr** [reply]
Yeah we need coupling between (potentially multiple) cell layers next to the boundaries, especially if we use Newton's method and want perfect Jacobians.

**Reply by lindsayad** [reply]
periodic enforcement in libMesh won't work for `DISCONTINUOUS` finite element types

**Reply by lindsayad** [reply]
we could make this work with a finite volume implementation of the `EqualValueConstraint`

**Reply by lindsayad** [reply]
but it would require adding more Lagrange multipliers to the system

**Reply by lindsayad** [reply]
(there is a penalty version we could use alternatively)

**Reply by roystgnr** [reply]
We do have periodic BCs on discontinuous variables; they ensure that the variable is allowed to be discontinuous from one side of the domain to the other!

More seriously: for a continuous FE type, u|left = u|right is the proper way to get a periodic variable u, but for a discontinuous type it's *not*.  You're at the very least coarsening all your boundary elements if you enforce a weak constraint strongly like that, and depending on your formulation you might also be hobbling your convergence rate.

The trouble is that we can't do weak periodic constriants at the libMesh level; the exact constraint term you want to add in those cases depends on your physics.
